numba/np/arrayobj.py

```diff
@@ -2569,95 +2571,143 @@ def array_flags_f_contiguous(context, builder, typ, value):
 # ------------------------------------------------------------------------------
 # .real / .imag
 
-@lower_getattr(types.Array, "real")
-def array_real_part(context, builder, typ, value):
+@overload_attribute(types.Array, 'real')
+def ol_array_real(arr):
+    typ = arr
     if typ.dtype in types.complex_domain:
-        return array_complex_attr(context, builder, typ, value, attr='real')
+        def impl(arr):
+            return _array_real_attr(arr)
+        return impl
     elif typ.dtype in types.number_domain:
-        # as an identity function
-        return impl_ret_borrowed(context, builder, typ, value)
+        def impl(arr):
+            return arr
+        return impl
     else:
-        raise NotImplementedError('unsupported .real for {}'.format(type.dtype))
+        msg = f"cannot access .real of array of {arr.dtype}"
+        raise errors.TypingError(msg)
 
 
-@lower_getattr(types.Array, "imag")
-def array_imag_part(context, builder, typ, value):
+@intrinsic
+def _force_readonly(tyctx, arr):
+    """Takes an array, and returns it with the readonly bit set in both typing
+    and lowering."""
+    sig = arr.copy(readonly=True)(arr)
+
+    def impl(cgctx, builder, sig, llargs):
+        arrayty = make_array(arr)
+        array = arrayty(cgctx, builder, llargs[0])
+        # make readonly hack
+        parent = array._datamodel.get_type('parent')
+        setattr(array, 'parent', Constant(cgctx.get_value_type(parent), None))
```

Comment:
I vaguely recall adding this with view of trying provide a "general" way to set the readonly bit, was wondering (here: https://github.com/numba/numba/pull/7999#discussion_r858960364) whether this ought to be a utility function? Am fine to remove it or move it into `numba.np.unsafe.ndarray`.

Reply:
Let's remove it from `_force_readonly`. I don't think this is the way to do it. The boxer is already setting the writeable flag according to the array readonly/mutable property: https://github.com/numba/numba/blob/9a7820cb5f21776587d23e7facf3b95fcbe6097d/numba/core/pythonapi.py#L1208. The "parent" affects the `OWNDATA` flag and it is a different concern.

Reply:
TBH, I'd rather that this didn't exist at all, but this sort of thing fails because the return type of the `impl` doesn't match the declared signature:
```python
from numba import njit
from numba.core.extending import overload
import numpy as np

def foo(x):
    pass

@overload(foo)
def ol_foo(x):
    def impl(x):
        return x
    return x.copy(readonly=True)(x), impl

@njit
def call_foo(x):
    return foo(x)

tmp = np.arange(5.)
print(call_foo(tmp))
```
so the intrinsic is needed to fix it. I'll go fix the patch WRT the above.